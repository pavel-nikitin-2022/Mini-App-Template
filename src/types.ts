export enum ProcessStatus {
  Start = 'Start',
  Pending = 'Pending',
  Result = 'Result',
}

export type FetchMLAnswerData = {
  animal: string
  probability: number
  coordinates: [number, number, number, number]
} | null

export type FetchMLAnswerResponse = {
  status: true
  data: FetchMLAnswerData
} | {
  status: false
  error: string
}

export type AnimalFormatter = {
  pattern: RegExp
  message: string
}

export type ServerAnswerFormatters = AnimalFormatter[]
