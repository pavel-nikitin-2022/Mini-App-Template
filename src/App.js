import React, { useState } from 'react'
import {
  View,
  ScreenSpinner,
  SplitLayout,
  SplitCol,
  Panel,
} from '@vkontakte/vkui'
import '@vkontakte/vkui/dist/vkui.css'

function App() {
  const [popout] = useState(<ScreenSpinner size="large" />)

  return (
    <SplitLayout popout={popout}>
      <SplitCol>
        <View>
          <Panel id="main" />
        </View>
      </SplitCol>
    </SplitLayout>
  )
}

export default App
