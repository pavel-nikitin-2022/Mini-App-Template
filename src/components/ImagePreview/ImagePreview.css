@keyframes scale {
  from {
    transform: scale(0);
  }

  to {
    transform: scale(1);
  }
}

.Preview {
  margin: 0 auto;
  display: inline-block;
  max-height: 270px;
  position: relative;
  backface-visibility: hidden;
  will-change: transform;
  transition: transform 0.75s ease-in-out;
}

.Preview_result {
  transform: translateY(-100px);
}

.Preview img {
  pointer-events: none;
  max-height: 100%;
  max-width: 100%;
  border-radius: 16px;
  -webkit-touch-callout: none;
  -webkit-user-select: none;
  -khtml-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.Preview_unload img {
  opacity: 0.5;
}

.Preview__spinner {
  position: absolute;
  z-index: 100;
}

.Preview__FoundObject {
  transform: scale(1);
  position: absolute;
  border: 2px solid green;
  transform: scale(0);
  animation: scale 0.75s linear 1s forwards;
}
