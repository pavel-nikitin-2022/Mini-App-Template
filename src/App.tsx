import React, { useCallback, useEffect, useRef, useState } from 'react'
import { useUploady } from '@rpldy/shared-ui'
import {
  DropPopup,
  ImagePreview,
  LoadButton,
  RestartButton,
  ResultCard,
} from './components'
import { SplitLayout, SplitCol, Panel, Placeholder } from '@vkontakte/vkui'
import { Icon56TagOutline } from '@vkontakte/icons'
import {
  useAbortAll,
  useItemFinishListener,
  useItemStartListener,
} from '@rpldy/uploady'
import { ProcessStatus, ServerAnswer } from './types'
import '@vkontakte/vkui/dist/vkui.css'
import './App.css'
import { serverParser } from './utils'

const App = (): JSX.Element => {
  const [isDragging, setIsDragging] = useState(false)
  const [processStatus, setProcessStatus] = useState(ProcessStatus.Start)
  const [imageWidth, setImageWidth] = React.useState(NaN)

  const serverAnswer = useRef<ServerAnswer>(null)

  const { upload } = useUploady()
  const abortAll = useAbortAll()

  const discardDrag = useCallback(() => setIsDragging(false), [])

  const abortConnection = useCallback(() => {
    abortAll()
    setProcessStatus(ProcessStatus.Start)
  }, [abortAll])

  const onRestart = useCallback(() => {
    setProcessStatus(ProcessStatus.Start)
    serverAnswer.current = null
  }, [])

  useItemFinishListener((item) => {
    serverAnswer.current = serverParser(item)
    setProcessStatus(ProcessStatus.Result)
  })

  useItemStartListener((event) => {
    const img = new Image()
    img.addEventListener('load', () => {
      window.URL.revokeObjectURL(img.src) // Free some memory
      setImageWidth(img.width)
    })
    img.src = window.URL.createObjectURL(event.file as unknown as Blob)
    setProcessStatus(ProcessStatus.Pending)
  })

  // Обработка события paste
  useEffect(() => {
    window.focus()
    function onPaste(e: ClipboardEvent) {
      const files = e.clipboardData?.files
      if (!files || processStatus !== ProcessStatus.Start) return
      upload(files as unknown as string[], { params: { test: 'paste' } })
    }
    window.addEventListener('paste', onPaste)

    return () => window.removeEventListener('paste', onPaste)
  }, [upload, processStatus])

  return (
    <SplitLayout>
      <SplitCol>
        <Panel
          onDragEnter={() => setIsDragging(true)}
          onMouseEnter={() => setIsDragging(false)}
        >
          <div className="App">
            <ImagePreview
              abortConnection={abortConnection}
              status={processStatus}
              imageWidth={imageWidth}
              coordinates={serverAnswer.current?.coordinates}
            />

            {processStatus === ProcessStatus.Result && (
              <ResultCard answer={serverAnswer.current} />
            )}

            {processStatus === ProcessStatus.Start && (
              <Placeholder
                icon={<Icon56TagOutline />}
                header="Определить питомца на фото"
                action={<LoadButton />}
              >
                Наши алгоритмы узнают, кто изображён на фотографии
              </Placeholder>
            )}

            {processStatus === ProcessStatus.Result && (
              <RestartButton onRestart={onRestart} />
            )}
          </div>

          {processStatus === ProcessStatus.Start && (
            <DropPopup {...{ discardDrag, isDraging: isDragging }} />
          )}
        </Panel>
      </SplitCol>
    </SplitLayout>
  )
}

export { App }
