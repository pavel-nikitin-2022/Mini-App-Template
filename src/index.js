import React from 'react'
import ReactDOM from 'react-dom'
import bridge from '@vkontakte/vk-bridge'
import { ConfigProvider, AdaptivityProvider, AppRoot } from '@vkontakte/vkui'
import App from './App'

bridge.send('VKWebAppInit')

ReactDOM.render(
  <ConfigProvider>
    <AdaptivityProvider>
      <AppRoot>
        <App />
      </AppRoot>
    </AdaptivityProvider>
  </ConfigProvider>,
  document.getElementById('root')
)
