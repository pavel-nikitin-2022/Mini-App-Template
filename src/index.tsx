import React from 'react'
import ReactDOM from 'react-dom'
import bridge from '@vkontakte/vk-bridge'
import Uploady from '@rpldy/uploady'
import { getMockSenderEnhancer } from '@rpldy/mock-sender'
import { ConfigProvider, AdaptivityProvider, AppRoot } from '@vkontakte/vkui'
import { fileFilter } from './utils'
import { App } from './App'

bridge.send('VKWebAppInit')
const mockSenderEnhancer = getMockSenderEnhancer({
  response: {
    multiobject_labels: [
      {
        status: 0,
        name: 'file',
        labels: [
          {
            eng: 'Person',
            rus: 'Человек',
            eng_categories: [],
            rus_categories: [],
            prob: 0.9586,
            coord: [100, 100, 300, 300],
          },
          {
            eng: 'Person',
            rus: 'Человек',
            eng_categories: [],
            rus_categories: [],
            prob: 0.9102,
            coord: [130, 325, 238, 428],
          },
          {
            eng: 'Person',
            rus: 'Человек',
            eng_categories: [],
            rus_categories: [],
            prob: 0.8765,
            coord: [208, 293, 258, 353],
          },
        ],
      },
    ],
  },
  delay: 3000,
})

ReactDOM.render(
  <ConfigProvider>
    <AdaptivityProvider>
      <AppRoot>
        <Uploady
          accept=".png,.jpg,.jpeg"
          fileFilter={fileFilter}
          multiple={false}
          clearPendingOnAdd
          maxConcurrent={1}
          enhancer={mockSenderEnhancer}
          destination={{
            url: 'https://lunacom.com',
            filesParamName: 'file',
            headers: {
              mode: ['scene', 'multiobject', 'pedestrian'],
              images: [
                {
                  name: 'file',
                },
              ],
            },
          }}
        >
          <App />
        </Uploady>
      </AppRoot>
    </AdaptivityProvider>
  </ConfigProvider>,
  document.getElementById('root')
)
