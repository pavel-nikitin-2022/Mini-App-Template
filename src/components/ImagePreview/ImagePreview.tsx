import UploadPreview from '@rpldy/upload-preview'
import React, { useEffect, useRef } from 'react'
import cx from 'classnames'
import { Button, Spinner } from '@vkontakte/vkui'
import { ProcessStatus } from '../../types'
import './ImagePreview.css'

type ImagePreviewProps = {
  status: ProcessStatus
  abortConnection: () => void
}

type PreviewMethods = {
  clear: () => void
  removePreview: (id: string) => void
}

const ImagePreview: React.FC<ImagePreviewProps> = ({
  status,
  abortConnection,
}) => {
  const previewMethodsRef = useRef<PreviewMethods>(null)
  const isLoading = status === ProcessStatus.Pending
  const isResult = status === ProcessStatus.Result

  useEffect(() => {
    if (status === ProcessStatus.Start && previewMethodsRef.current)
      previewMethodsRef.current.clear()
  }, [status])

  return (
    <>
      <div
        className={cx('Preview', {
          Preview_unload: isLoading,
          Preview_result: isResult,
        })}
      >
        {isLoading && <Spinner className="Preview__spinner" size="large" />}

        <UploadPreview previewMethodsRef={previewMethodsRef} />
      </div>

      {isLoading && (
        <Button
          size="l"
          mode="link"
          onClick={() => {
            abortConnection()
            if (!previewMethodsRef.current) return
            previewMethodsRef.current.clear()
          }}
        >
          Отмена
        </Button>
      )}
    </>
  )
}

export default React.memo(ImagePreview)
