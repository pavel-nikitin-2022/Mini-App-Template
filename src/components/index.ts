import { ImagePreview } from './ImagePreview/ImagePreview'
import { LoadButton } from './LoadButton/LoadButton'
import { DropPopup } from './DropPopup/DropPopup'
import { ResultCard } from './ResultCard/ResultCard'
import { RestartButton } from './RestartButton/RestartButton'

export { ImagePreview, LoadButton, DropPopup, ResultCard, RestartButton }
