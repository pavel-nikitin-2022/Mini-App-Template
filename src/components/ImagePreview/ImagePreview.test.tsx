import React from 'react'
import { fireEvent, render, screen } from '@testing-library/react'
import { ImagePreview, ImagePreviewProps } from './ImagePreview'
import Uploady from '@rpldy/uploady'
import { ProcessStatus } from 'src/types'

function setUp(props: Partial<ImagePreviewProps>) {
  return render(
    <Uploady>
      <ImagePreview
        abortConnection={() =>
          props.abortConnection ? props.abortConnection() : null
        }
        imageWidth={props.imageWidth ?? 1}
        status={props.status ?? ProcessStatus.Start}
        coordinates={props.coordinates ?? [1, 2, 3, 4]}
      />
    </Uploady>
  )
}

describe('ImagePreview', () => {
  it('Проверка интерфейса при status start', () => {
    const { asFragment } = setUp({ status: ProcessStatus.Start })
    expect(asFragment).toMatchSnapshot()
  })

  it('Проверка интерфейса при status pending', () => {
    const { asFragment } = setUp({ status: ProcessStatus.Pending })
    expect(asFragment).toMatchSnapshot()
  })

  it('Проверка интерфейса при status result', () => {
    const { asFragment } = setUp({ status: ProcessStatus.Result })
    expect(asFragment).toMatchSnapshot()
  })

  it('Проверка кнопки отмены при загрузке ответа', () => {
    const abortConnection = jest.fn()
    setUp({
      status: ProcessStatus.Pending,
      abortConnection: abortConnection,
    })
    const element = screen.getByRole('button')
    fireEvent.click(element)
    expect(abortConnection).toHaveBeenCalledTimes(1)
  })
})
