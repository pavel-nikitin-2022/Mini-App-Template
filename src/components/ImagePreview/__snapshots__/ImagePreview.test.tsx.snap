// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImagePreview Проверка интерфейса при status pending 1`] = `[Function]`;

exports[`ImagePreview Проверка интерфейса при status result 1`] = `[Function]`;

exports[`ImagePreview Проверка интерфейса при status start 1`] = `[Function]`;
